### DISCUSSION TITLE
Large trig file

### DISCUSSION THREAD

**Discussion by RichDijk** [oxigraph/oxigraph#445] [Q&A]
Hi,

I am trying to read a 235 GB trig file into a 32 GB Windows machine using oxigraph-server, but do get a memory allocation error.
Should this be possible or do I ask to much from the resources?

regards, Richard

**Comment by Tpt** [comment]
Hi! Sorry for that. It should be possible to load files of similar size and I already managed to do that. It's likely an error in the code that guess how big the batches can be before overloading the memory. It's a bit of a "wet finger" estimation. Do you have a backtrace of the error? Are you using a slow disk by chance? How many cores your machine has?

**Reply by RichDijk** [reply]
I am using an11th gen i7-11850H having 8 cores with 500GB solid state disk, feels sufficient.
The error I get is 

` memory allocation of 53687091200 bytes failed `

This message is the same for both using oxigraph-server and pyoxigraph.

**Reply by Tpt** [reply]
Thank you! This malloc call seems way to big. There must be a bug somewhere.

**Reply by RichDijk** [reply]
Could this also be a diskspace issue? I do miss about 52GB for making a copy of the trig source.
oxigraph was not writing to disk yet, but the calculations may lead to this message.

**Reply by Tpt** [reply]
That's definitely possible. But it would be strange, Oxigraph should not allocate such large files (except if you have very long IRIs or literals).

**Reply by Tpt** [reply]
Quick question I forgot to ask: are you using `oxigraph_server load` or are you doing an HTTP request?

**Reply by RichDijk** [reply]
I am doing a load of the trig file creating a store.
So a load file using server and bulk_load using pyoxigraph.

**Reply by Tpt** [reply]
I have made a change in the hope to limit the memory consumption (I avoid trying to autodetect system parameter and I fallback to safe defaults): #446 

I have a slight hope it will work better now.

Server binaries and Python wheels with this change are availlable here in the "Artifacts" section: https://github.com/oxigraph/oxigraph/actions/runs/4482484952

> Could this also be a diskspace issue? I do miss about 52GB for making a copy of the trig source.

That's probably gonna be a problem during the load. Oxigraph is sadly quite memory hungry at the moment.

**Reply by RichDijk** [reply]
Sorry, but no success. Error message is still, even after releasing some diskspace. 
trig file has size 234GB , free diskspace is 145GB

Executing:
`oxigraph_server_patch.exe --location ./store load --file staging_store.trig`

Gives:
`memory allocation of 53687091200 bytes failed`

**Reply by Tpt** [reply]
Sorry again for that. I have made a PR that adds more verbose logging. Here are the artifacts: https://github.com/oxigraph/oxigraph/actions/runs/4493365449?pr=447 I hope it might help. I would be very curious to see at which point the loader fails.

**Reply by RichDijk** [reply]
A new run gives

```
Starting loading with 2 threads and a batch size of 1000000
memory allocation of 53687091200 bytes failed
 ```

**Reply by Tpt** [reply]
Thank you! Is it when called from the command line? Are you 100% sure your TriG file is correct?

**Reply by RichDijk** [reply]
yes, from command line. 
100% sure, not sure... File is created by Virtuoso. I am not aware of another method for checking this. Do you?

**Reply by Tpt** [reply]
> File is created by Virtuoso

Thank you! So, should be correct. To validate it you might try to parse it with the `parse` function of pyoxigraph. It will also help me to narrow the bug to a parser bug (if the `parse` function fails with an out-of-memory error) or a loader bug (if `parse` works correctly). I am so sorry for not being more helpful.

**Reply by RichDijk** [reply]
I think it is a parser error because (using the original pyoxigraph version; so not one of the 2 latest)

```
In [1]: import pyoxigraph as pox

In [2]: p = pox.parse('staging_store.trig', 'application/trig')

In [3]: list(p)[0]
memory allocation of 53687091200 bytes failed

In [3]: pox.__version__
Out[3]: '0.3.14'
```

Using pyoxigraph with a smaller trig file from the same triplestore has no issues.

**Reply by Tpt** [reply]
Great! We are moving forward. Thank you!

I wrote [a patch for the Turtle parser](https://github.com/oxigraph/oxigraph/actions/runs/4505107390) to limit the parser buffers to 10Mb. I hope it might help narrow the problem. [Artifacts with the patch should be available soon here](https://github.com/oxigraph/oxigraph/actions/runs/4505107390).

**Reply by RichDijk** [reply]
Sorry to say but both the server exe and wheel do give the same memory error.

**Reply by Tpt** [reply]
Sorry for that. Would you be able to share me a snippet of the file by chance to help me reproduce the bug? The first 2M triples should be fine.

**Reply by RichDijk** [reply]
This is file with the first 10000 lines
[trip_part.txt](https://github.com/oxigraph/oxigraph/files/11076434/trip_part.txt)

**Reply by Tpt** [reply]
Thank you! I try to execute the parsing of this file concatenated 10,000 times. The memory consumption stayed stable at less than 10MB. The bug is maybe related to a later position in the file.

**Reply by RichDijk** [reply]
Trying to count the triples I ran:
```
with open('staging_store.trig','r') as fp:
    for count, line in enumerate(fp):
        pass
print('Total Lines', count + 1)
```
gave
```
Traceback (most recent call last):
  File "c:\data\https___data.labs.kadaster.nl_staging_bag-model_graphs_default-14.trig\lees2mb.py", line 18, in <module>
    for count, line in enumerate(fp):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.2800.0_x64__qbz5n2kfra8p0\lib\encodings\cp1252.py", line 23, in decode    
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x8d in position 2361: character maps to <undefined>
``` 
Sorry for this thread with this imperfect file but maybe this issue is handled incorrectly by the oxiserver I use.

**Reply by Tpt** [reply]
Sorry for the answer lag. I am not sure this error is related to the server bug because if the file encoding was wrong it is likely it should also fail on Linux. `open('staging_store.trig','r')` is decoding the text and it seems it is not usinf utf8 but cp1252. Is it still failing with `open('staging_store.trig','rb')` to give the binary data to Oxigraph? Or `open('staging_store.trig','rt', encoding="utf8")`?
